File: ts/src/render.ts
/**
 * Display formatting (the spec's display forms) and Mustache rendering with
 * the contract render context — see the Go reference render.go. The context
 * is part of the cross-language contract: fields / groups / values with
 * precomputed helper forms; inactive fields are excluded.
 */

import { isMessage } from "@bufbuild/protobuf";
import { DurationSchema, TimestampSchema } from "@bufbuild/protobuf/wkt";
import { formatGoDuration, formatRfc3339 } from "./duration.js";
import type { Schema_Field } from "./gen/schemapb/schema_pb.js";
import { asBigInt, isNativeStruct, type Native, type NativeStruct, toNative } from "./value.js";

/**
 * Renders a native value in the spec's display form: integers in decimal,
 * doubles in JSON number form, bool as true/false, durations in Go form
 * ("5m0s"), timestamps as RFC3339, bytes as std base64, lists and objects as
 * compact JSON with the same leaf forms.
 */
export function displayString(v: Native): string {
  if (v === null) {
    return "";
  }
  if (typeof v === "string") {
    return v;
  }
  if (typeof v === "boolean") {
    return v ? "true" : "false";
  }
  if (typeof v === "bigint") {
    return v.toString(10);
  }
  if (typeof v === "number") {
    return JSON.stringify(v);
  }
  if (v instanceof Uint8Array) {
    return base64(v);
  }
  if (isMessage(v, DurationSchema)) {
    return formatGoDuration(v);
  }
  if (isMessage(v, TimestampSchema)) {
    return formatRfc3339(v);
  }
  return JSON.stringify(displayJson(v));
}

/**
 * Converts non-JSON leaves (duration, timestamp, bytes, bigint) to their
 * display forms so container JSON stays in the spec's leaf forms. Object
 * keys are sorted — Go marshals maps sorted.
 */
function displayJson(v: Native): unknown {
  if (v === null || typeof v === "boolean" || typeof v === "number" || typeof v === "string") {
    return v;
  }
  if (typeof v === "bigint") {
    // Go marshals int64 as a JSON number; stay exact within double range.
    return v >= BigInt(Number.MIN_SAFE_INTEGER) && v <= BigInt(Number.MAX_SAFE_INTEGER)
      ? Number(v)
      : v.toString(10);
  }
  if (v instanceof Uint8Array || isMessage(v, DurationSchema) || isMessage(v, TimestampSchema)) {
    return displayString(v);
  }
  if (Array.isArray(v)) {
    return v.map(displayJson);
  }
  const out: Record<string, unknown> = {};
  for (const key of Object.keys(v).sort()) {
    const el = v[key];
    if (el !== undefined) {
      out[key] = displayJson(el);
    }
  }
  return out;
}

const BASE64_CHARS = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

/** Std base64 without Node Buffer (portable to browsers). */
export function base64(bytes: Uint8Array): string {
  let out = "";
  for (let i = 0; i < bytes.length; i += 3) {
    const b0 = bytes[i] ?? 0;
    const b1 = bytes[i + 1];
    const b2 = bytes[i + 2];
    out += BASE64_CHARS[b0 >> 2];
    out += BASE64_CHARS[((b0 & 3) << 4) | ((b1 ?? 0) >> 4)];
    out += b1 === undefined ? "=" : BASE64_CHARS[((b1 & 15) << 2) | ((b2 ?? 0) >> 6)];
    out += b2 === undefined ? "=" : BASE64_CHARS[b2 & 63];
  }
  return out;
}

/** Decodes std base64; undefined on invalid input. */
export function base64Decode(s: string): Uint8Array | undefined {
  if (!/^[A-Za-z0-9+/]*={0,2}$/.test(s) || s.length % 4 !== 0) {
    return undefined;
  }
  const clean = s.replace(/=+$/, "");
  const out: number[] = [];
  let bits = 0;
  let acc = 0;
  for (const ch of clean) {
    acc = (acc << 6) | BASE64_CHARS.indexOf(ch);
    bits += 6;
    if (bits >= 8) {
      bits -= 8;
      out.push((acc >> bits) & 0xff);
    }
  }
  return new Uint8Array(out);
}

// ASCII-only casing: locale-dependent Unicode casing would break
// cross-implementation determinism.
export function asciiUpper(s: string): string {
  return s.replace(/[a-z]/g, (c) => String.fromCharCode(c.charCodeAt(0) - 32));
}

export function asciiLower(s: string): string {
  return s.replace(/[A-Z]/g, (c) => String.fromCharCode(c.charCodeAt(0) + 32));
}

/** Go strconv.Quote-compatible quoting for the render context. */
export function goQuote(s: string): string {
  return JSON.stringify(s);
}

/** The short kind name used in render contexts. */
export function kindName(f: Schema_Field): string {
  switch (f.kind.case) {
    case "float":
      return "float";
    case "double":
      return "double";
    case "int32":
      return "int32";
    case "int64":
      return "int64";
    case "uint32":
      return "uint32";
    case "uint64":
      return "uint64";
    case "bool":
      return "bool";
    case "string":
      return "string";
    case "bytes":
      return "bytes";
    case "choice":
      return "choice";
    case "duration":
      return "duration";
    case "timestamp":
      return "timestamp";
    case "list":
      return "list";
    case "object":
      return "object";
    case "map":
      return "map";
    case "oneOf":
      return "oneof";
    case "ref":
      return "ref";
    case "computed":
      return "computed";
    case "json":
      return "json";
    default:
      return "";
  }
}

/** One field entry of the contract render context. */
export interface RenderField {
  name: string;
  title: string;
  description: string;
  unit: string;
  group: string;
  kind: string;
  label: string;
  set: boolean;
  computed: boolean;
  secret: boolean;
  immutable: boolean;
  deprecated: boolean;
  value: string;
  onoff: string;
  yesno: string;
  quoted: string;
  upper: string;
  lower: string;
}

export interface RenderContext {
  fields: RenderField[];
  groups: { name: string; fields: RenderField[] }[];
  values: Record<string, string>;
}

/** Builds one render-context field entry (label from a matched choice option). */
export function renderField(f: Schema_Field, values: NativeStruct): RenderField {
  const val = values[f.name];
  const set = f.name in values && val !== null && val !== undefined;
  const display = set ? displayString(val ?? null) : "";
  let label = "";
  if (f.kind.case === "choice" && set) {
    for (const o of f.kind.value.options) {
      if (nativeEquals(val ?? null, toNative(o.value))) {
        label = o.label;
        break;
      }
    }
  }
  const b = typeof val === "boolean" ? val : false;
  return {
    name: f.name,
    title: f.title ?? "",
    description: f.description ?? "",
    unit: f.unit ?? "",
    group: f.group ?? "",
    kind: kindName(f),
    label,
    set,
    computed: f.kind.case === "computed",
    secret: f.secret,
    immutable: f.immutable,
    deprecated: f.deprecated,
    value: display,
    onoff: b ? "on" : "off",
    yesno: b ? "yes" : "no",
    quoted: goQuote(display),
    upper: asciiUpper(display),
    lower: asciiLower(display),
  };
}

/** Structural native equality with cross-numeric comparison (spec). */
export function nativeEquals(a: Native, b: Native): boolean {
  const an = numericOf(a);
  if (an !== undefined) {
    const bn = numericOf(b);
    return bn !== undefined && an === bn;
  }
  if (a === null || b === null) {
    return a === b;
  }
  if (typeof a === "boolean" || typeof a === "string") {
    return a === b;
  }
  if (a instanceof Uint8Array) {
    return b instanceof Uint8Array && a.length === b.length && a.every((x, i) => b[i] === x);
  }
  if (isMessage(a, DurationSchema)) {
    return isMessage(b, DurationSchema) && a.seconds === b.seconds && a.nanos === b.nanos;
  }
  if (isMessage(a, TimestampSchema)) {
    return isMessage(b, TimestampSchema) && a.seconds === b.seconds && a.nanos === b.nanos;
  }
  if (Array.isArray(a)) {
    return (
      Array.isArray(b) && a.length === b.length && a.every((x, i) => nativeEquals(x, b[i] ?? null))
    );
  }
  if (isNativeStruct(a) && isNativeStruct(b)) {
    const ak = Object.keys(a);
    const bk = Object.keys(b);
    return (
      ak.length === bk.length && ak.every((k) => k in b && nativeEquals(a[k] ?? null, b[k] ?? null))
    );
  }
  return false;
}

/** Numeric view for cross-type equality: exact when both are integral. */
function numericOf(v: Native): number | bigint | undefined {
  if (typeof v === "bigint") {
    return v;
  }
  if (typeof v === "number") {
    const i = asBigInt(v);
    return i ?? v;
  }
  return undefined;
}

File: rust/src/render.rs
//! Display formatting (the spec's display forms), native equality and the
//! contract render context entries.

use base64::Engine as _;
use serde_json::json;

use crate::duration::{format_go_duration, format_rfc3339};
use crate::value::{as_double, as_int, to_native, Native, NativeStruct, SchemaField};

/// The spec's display form of a native value.
#[must_use]
pub fn display_string(v: &Native) -> String {
    match v {
        Native::Null => String::new(),
        Native::Str(s) => s.clone(),
        Native::Bool(b) => if *b { "true" } else { "false" }.to_owned(),
        Native::Int(n) => n.to_string(),
        Native::Uint(n) => n.to_string(),
        Native::Double(_) | Native::List(_) | Native::Struct(_) => {
            serde_json::to_string(&display_json(v)).unwrap_or_default()
        }
        Native::Bytes(b) => base64::engine::general_purpose::STANDARD.encode(b),
        Native::Duration(d) => format_go_duration(d),
        Native::Timestamp(t) => format_rfc3339(t),
    }
}

/// Container JSON with the spec's leaf forms (sorted keys via `BTreeMap`).
fn display_json(v: &Native) -> serde_json::Value {
    #[allow(clippy::cast_precision_loss, clippy::cast_possible_truncation)]
    match v {
        Native::Null => serde_json::Value::Null,
        Native::Bool(b) => json!(b),
        Native::Int(n) => json!(n),
        Native::Uint(n) => json!(n),
        Native::Double(f) => {
            // Go's encoding/json renders integral doubles without ".0".
            if f.fract() == 0.0 && f.is_finite() && f.abs() < 1e15 {
                json!(*f as i64)
            } else {
                json!(f)
            }
        }
        Native::Str(s) => json!(s),
        Native::Bytes(_) | Native::Duration(_) | Native::Timestamp(_) => {
            json!(display_string(v))
        }
        Native::List(items) => serde_json::Value::Array(items.iter().map(display_json).collect()),
        Native::Struct(m) => serde_json::Value::Object(
            m.iter()
                .map(|(k, x)| (k.clone(), display_json(x)))
                .collect(),
        ),
    }
}

/// ASCII-only casing: locale-dependent Unicode casing would break
/// cross-implementation determinism.
#[must_use]
pub fn ascii_upper(s: &str) -> String {
    s.chars()
        .map(|c| {
            if c.is_ascii_lowercase() {
                c.to_ascii_uppercase()
            } else {
                c
            }
        })
        .collect()
}

#[must_use]
pub fn ascii_lower(s: &str) -> String {
    s.chars()
        .map(|c| {
            if c.is_ascii_uppercase() {
                c.to_ascii_lowercase()
            } else {
                c
            }
        })
        .collect()
}

/// Go `strconv.Quote`-compatible quoting for the render context.
#[must_use]
pub fn go_quote(s: &str) -> String {
    serde_json::to_string(s).unwrap_or_default()
}

/// The short kind name used in render contexts.
#[must_use]
pub const fn kind_name(f: &SchemaField) -> &'static str {
    use crate::gen::schemapb::schema::field::Kind as K;
    match f.kind.as_ref() {
        Some(K::Float(_)) => "float",
        Some(K::Double(_)) => "double",
        Some(K::Int32(_)) => "int32",
        Some(K::Int64(_)) => "int64",
        Some(K::Uint32(_)) => "uint32",
        Some(K::Uint64(_)) => "uint64",
        Some(K::Bool(_)) => "bool",
        Some(K::String(_)) => "string",
        Some(K::Bytes(_)) => "bytes",
        Some(K::Choice(_)) => "choice",
        Some(K::Duration(_)) => "duration",
        Some(K::Timestamp(_)) => "timestamp",
        Some(K::List(_)) => "list",
        Some(K::Object(_)) => "object",
        Some(K::Map(_)) => "map",
        Some(K::OneOf(_)) => "oneof",
        Some(K::Ref(_)) => "ref",
        Some(K::Computed(_)) => "computed",
        Some(K::Json(_)) => "json",
        None => "",
    }
}

/// Structural native equality with cross-numeric comparison (spec).
#[must_use]
pub fn native_equals(a: &Native, b: &Native) -> bool {
    if let (Some(x), Some(y)) = (numeric_int(a), numeric_int(b)) {
        return x == y;
    }
    if matches!(a, Native::Int(_) | Native::Uint(_) | Native::Double(_))
        || matches!(b, Native::Int(_) | Native::Uint(_) | Native::Double(_))
    {
        return match (as_double(a), as_double(b)) {
            (Some(x), Some(y)) => (x - y).abs() == 0.0,
            _ => false,
        };
    }
    match (a, b) {
        (Native::List(x), Native::List(y)) => {
            x.len() == y.len() && x.iter().zip(y).all(|(l, r)| native_equals(l, r))
        }
        (Native::Struct(x), Native::Struct(y)) => {
            x.len() == y.len()
                && x.iter()
                    .all(|(k, l)| y.get(k).is_some_and(|r| native_equals(l, r)))
        }
        _ => a == b,
    }
}

fn numeric_int(v: &Native) -> Option<i128> {
    match v {
        Native::Int(n) => Some(i128::from(*n)),
        Native::Uint(n) => Some(i128::from(*n)),
        Native::Double(_) => as_int(v).map(i128::from),
        _ => None,
    }
}

/// One field entry of the contract render context (serialized for Mustache).
#[derive(Debug, Clone, serde::Serialize)]
pub struct RenderField {
    pub name: String,
    pub title: String,
    pub description: String,
    pub unit: String,
    pub group: String,
    pub kind: &'static str,
    pub label: String,
    pub set: bool,
    pub computed: bool,
    pub secret: bool,
    pub immutable: bool,
    pub deprecated: bool,
    pub value: String,
    pub onoff: &'static str,
    pub yesno: &'static str,
    pub quoted: String,
    pub upper: String,
    pub lower: String,
}

#[derive(Debug, Clone, serde::Serialize)]
pub struct RenderGroup {
    pub name: String,
    pub fields: Vec<RenderField>,
}

#[derive(Debug, Clone, serde::Serialize)]
pub struct RenderContext {
    pub fields: Vec<RenderField>,
    pub groups: Vec<RenderGroup>,
    pub values: std::collections::BTreeMap<String, String>,
}

/// Builds one render-context field entry.
#[must_use]
pub fn render_field(f: &SchemaField, values: &NativeStruct) -> RenderField {
    let val = values.get(&f.name);
    let set = val.is_some_and(|v| !v.is_null());
    let display = if set {
        val.map(display_string).unwrap_or_default()
    } else {
        String::new()
    };
    let mut label = String::new();
    if set {
        if let Some(crate::gen::schemapb::schema::field::Kind::Choice(ch)) = f.kind.as_ref() {
            for o in &ch.options {
                if val.is_some_and(|v| native_equals(v, &to_native(o.value.as_ref()))) {
                    label.clone_from(&o.label);
                    break;
                }
            }
        }
    }
    let b = matches!(val, Some(Native::Bool(true)));
    RenderField {
        name: f.name.clone(),
        title: f.title.clone().unwrap_or_default(),
        description: f.description.clone().unwrap_or_default(),
        unit: f.unit.clone().unwrap_or_default(),
        group: f.group.clone().unwrap_or_default(),
        kind: kind_name(f),
        label,
        set,
        computed: matches!(
            f.kind.as_ref(),
            Some(crate::gen::schemapb::schema::field::Kind::Computed(_))
        ),
        secret: f.secret,
        immutable: f.immutable,
        deprecated: f.deprecated,
        quoted: go_quote(&display),
        upper: ascii_upper(&display),
        lower: ascii_lower(&display),
        value: display,
        onoff: if b { "on" } else { "off" },
        yesno: if b { "yes" } else { "no" },
    }
}
